### PR TITLE
DB Storage Indexing & Error Handling 

### LINKED ISSUE #13: Add Database Storage for Clip Interaction and Message IDs
## Description
We need to add functionality to our database to store the interaction IDs and message IDs for each clip submitted. This will allow the bot to interact with these clips later, such as updating ratings or managing the weekly competition.

## Objectives
- [ ] Design database schema to include fields for interaction ID and message ID
- [ ] Implement function to save interaction ID and message ID when a clip is submitted
- [ ] Create function to retrieve clip information by interaction ID or message ID
- [ ] Add error handling for database operations (try catches)

## Technical Details
- Database: MongoDB
- Relevant files:
  - `src/database/db.js`

## Acceptance Criteria
- [ ] Database schema includes fields for `interactionId` and `messageId`
- [ ] Clip submission process successfully stores both IDs in the database
- [ ] Bot can retrieve clip information using either ID
- [ ] All database operations have appropriate error handling
- [ ] Existing functionality (if any) is not broken by these changes

## Additional Notes
- Consider indexing the new ID fields for faster lookups
- Ensure that the database can handle Discord's snowflake IDs (64-bit integers)

### PR DESCRIPTION
**Pull request by RaudrSkol**
# Add Database Storage and Indexing

## Description
This pull request implements database storage and indexing for the Clip Rating Bot, enhancing its data persistence capabilities and query performance.

### 1. Database Schema Updates
- Updated the Clip schema to include indexing for frequently queried fields.
- Added new fields to support more efficient clip retrieval and sorting.

### 2. Database Operations
- Implemented new database operations in `src/database/db.js` for improved clip storage and retrieval.
- Added methods for efficient querying of clips based on various criteria (e.g., submission date, rating).

### 3. Indexing Strategy
- Created indexes on the `submittedBy`, `createdAt`, and `averageRating` fields to optimize query performance.
- Implemented a compound index for weekly competition queries.

### 4. Query Optimization
- Refactored existing database queries to utilize the new indexes.
- Implemented pagination for clip retrieval to handle large datasets efficiently.

### CODE REVIEW

File: src/database/db.js [hunk 1]
```diff
@@ -34,56 +36,120 @@ const Clip = mongoose.model('Clip', ClipSchema);
...
-      clipId,
-      { $push: { ratings: { rating, ratedBy } } },
-      { new: true }
-    );
+    try {
+      return await Clip.findByIdAndUpdate(
+        clipId,
+        { $push: { ratings: { rating, ratedBy } } },
+        { new: true }
+      );
```
Comment (jSydorowicz21) [inline src/database/db.js]: Does this allow users to update ratings, or only assign new ones?

File: src/database/db.js [hunk 2]
```diff
@@ -34,56 +36,120 @@ const Clip = mongoose.model('Clip', ClipSchema);
...
+  },
+
+  getClipByInteractionId: async (interactionId) => {
+    try {
+      return await Clip.findOne({ interactionId });
+    } catch (error) {
+      console.error('Error getting clip by interaction ID:', error);
+      throw error;
+    }
+  },
```
Comment (jSydorowicz21) [inline src/database/db.js]: This schema will attempt to find the clip by comparing your interactionId that you provided with the mongoDB unique generated id. Should define your query similar to how you did  with get clips by user

File: src/database/db.js [hunk 3]
```diff
@@ -34,56 +36,120 @@ const Clip = mongoose.model('Clip', ClipSchema);
...
+
+  getClipByMessageId: async (messageId) => {
+    try {
+      return await Clip.findOne({ messageId });
+    } catch (error) {
+      console.error('Error getting clip by message ID:', error);
+      throw error;
+    }
```
Comment (jSydorowicz21) [inline src/database/db.js]: Same with this one

File: src/index.js [hunk 4]
```diff
@@ -60,7 +67,7 @@ connectDB().then(async () => {
 });
 
 // Now you can use db operations in your bot commands, for example:
-// db.storeClip(url, description, submittedBy)
+// db.storeClip(url, description, submittedBy, interactionId, messageId)
 // db.addRating(clipId, rating, ratedBy)
 // db.getAllClips()
 // etc.
```
Comment (jSydorowicz21) [inline src/index.js]: remove extra comments

**Review by jSydorowicz21 — CHANGES_REQUESTED** [review]
Small fixes and need a squash on commits

File: src/database/db.js (hunk 1 above)
Comment (RaudrSkol) [reply src/database/db.js]: This only allows users to add new ratings:

**Key points:**
1. It always adds a new rating object to the array.
2. It doesn't check if the user (ratedBy) has already rated the clip.
3. There's no mechanism to update an existing rating.

Should we update this to allow updates?

Comment (jSydorowicz21) [reply src/database/db.js]: Yes, either that or add a new method to update ratings as we should expect users to be able to change their ratings

**Review by jSydorowicz21 — APPROVED** [review]

**Pull request merged**